src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/HistogramAdvice.cs

```diff
@@ -0,0 +1,71 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Contains settings used to advise metrics consumers how to construct storage for <see cref="Histogram{T}"/> instruments.
+    /// </summary>
+    /// <typeparam name="T">Histogram value type.</typeparam>
+    public sealed class HistogramAdvice<T> where T : struct
```

Comment:
What's the plan for https://github.com/open-telemetry/opentelemetry-specification/blob/6b5e9404a3c574c1239230d0753665b4add0b977/specification/metrics/api.md?plain=1#L272 which is coming? Is there a way that .NET doesn't have to keep changing as more advisory parameters got added in the spec? Is there a way that .NET can provide some infrastructure so OpenTelemetry .NET can extend and experiment without waiting for .NET runtime vNext?

Reply:
We discussed this a bit on the proposal. I think introducing a base class in the future could work. Check out: https://github.com/dotnet/runtime/issues/63650#issuecomment-2123396445

We didn't discuss making it extensible 🤔

Reply:
Here is my general thinking: OpenTelemetry will keep adding more advisory parameters, it will be a downside if .NET is playing the catch-up game based on 12 months release frequency. I suggest that .NET has a generic solution which allows arbitrary advisory parameters to be passed from the instrumentation author to the SDK, so OpenTelemetry SDK can experiment with new things without having to wait for another year.

Reply:
What exactly are you suggesting @reyang? Do you suggest enabling this for all instruments? and allow adding any arbitrary parameters? There was discussion adding attributes to the advisory, but this is experimental.

Reply:
> Do you suggest enabling this for all instruments?

Yes. Allow `object` to be passed in, so OpenTelemetry .NET can try new things without having to wait for another year.

Reply:
Strongly typed advisory parameter can be considered as a plus (e.g. static check).

Reply:
> Yes. Allow object to be passed in, so OpenTelemetry .NET can try new things without having to wait for another year.

I don't think this will work well. We need the instrument publishers to provide that information. Without specifying specific types/values, publishers won't be able to use it, especially if changes are needed in the future.

Reply:
I would guess we'll see 10-1000x more usage of officially supported advice vs. experimental advice over time so I certainly would be reticent to degrade the experience of supported advice as part of making experimental advice possible. Changing a strong type like HistogramAdvice<T> into object would degrade that experience because now the IDE wouldn't offer any suggestion about what parameter type is appropriate.

A couple possibilities that I think would not degrade the experience for supported advice:
1. OTel could define extension methods and a ConditionalWeakTable<> to attach arbitrary state to an instrument. The downside is that any code using the experimental advice must have an OTel assembly reference. If we think its primarily app developers who will use experimental stuff thats probably not a problem but library developers its likely to discourage them.
2. We could add some API that puts a property bag on the Advice object, or directly on the Instrument object. This is similar to the AddCustomProperty() API on Activity.
3. In theory we could abuse instrument tags as a general purpose property bag but all non OTel consumers would probably misinterpret experimental advice as some kind of instrument dimension which would look weird.